// carr-backend/src/app.ts
import express from 'express';
import helmet from 'helmet';
import cors from 'cors';
import cookieParser from 'cookie-parser';
import { env } from './config/env';
import { logger } from './config/logger';
import { connectDatabase } from './database/connection';
import { globalLimiter } from './middleware/rateLimiter';
import { errorHandler, notFoundHandler } from './middleware/error';
import routes from './routes';
import cron from 'node-cron';
import { ReminderJob } from './jobs/reminder.job';
import { createProxyMiddleware } from 'http-proxy-middleware';

// Initialize Redis (auto-connects on import)
import './config/redis';

const app = express();

// ─── Security ────────────────────────────────────────────
app.use(helmet());
app.use(
  cors({
    origin: (origin, callback) => {
      // Allow requests with no origin (like mobile apps or curl requests)
      if (!origin) return callback(null, true);

      const allowedOrigins = [
        env.CLIENT_URL,
        'http://localhost:3000',
        'http://localhost:3001',
        'http://127.0.0.1:3000',
        'http://127.0.0.1:3001',
        'https://uniexo.in'
      ];

      // Allow any vercel domain automatically
      if (allowedOrigins.includes(origin) || origin.endsWith('.vercel.app')) {
        callback(null, true);
      } else {
        callback(new Error('Not allowed by CORS'));
      }
    },
    credentials: true,
    methods: ['GET', 'POST', 'PUT', 'PATCH', 'DELETE', 'OPTIONS'],
    allowedHeaders: ['Content-Type', 'Authorization'],
  }),
);

// ─── Parsers ─────────────────────────────────────────────
app.use(express.json({ limit: '10mb' }));
app.use(express.urlencoded({ extended: true, limit: '10mb' }));
app.use(cookieParser());

// ─── Rate Limiting ───────────────────────────────────────
app.use(globalLimiter);

// ─── Health Check ────────────────────────────────────────
app.get('/health', (_req, res) => {
  res.status(200).json({
    status: 'ok',
    timestamp: new Date().toISOString(),
    uptime: process.uptime(),
  });
});

app.use('/api/v1', routes);

// ─── Frontend Proxy (Integrated Mode) ────────────────────
// If we're not on Vercel and in production, proxy all other requests to the frontend
if (!process.env.VERCEL) {
  app.use(
    '/',
    createProxyMiddleware({
      target: 'http://localhost:3001',
      changeOrigin: true,
      ws: true, // support websockets for hot reloading in dev
      logLevel: 'error',
      // Don't proxy if it's an API route or health check (already handled above)
      filter: (pathname) => {
        return !pathname.startsWith('/api') && !pathname.startsWith('/health');
      },
    }),
  );
}

// ─── Scheduled Jobs ──────────────────────────────────────
// Run every day at midnight
cron.schedule('0 0 * * *', () => {
  ReminderJob.run();
});

// ─── Error Handling ──────────────────────────────────────
app.use(notFoundHandler);
app.use(errorHandler);

// ─── Start Server ────────────────────────────────────────
const startServer = async () => {
  try {
    await connectDatabase();

    // Only listen if not running on Vercel
    if (!process.env.VERCEL) {
      app.listen(env.PORT, () => {
        logger.info(`🚀 Server running on port ${env.PORT} in ${env.NODE_ENV} mode`);
      });
    } else {
      logger.info('🚀 App initialized for Vercel Serverless environment');
    }
  } catch (error) {
    logger.error('Failed to start server:', error);
    if (!process.env.VERCEL) {
      process.exit(1);
    }
  }
};

// For Vercel, we need to ensure the database is connected
if (process.env.VERCEL) {
  connectDatabase().catch(err => logger.error('Vercel DB Connection Error:', err));
} else {
  startServer();
}

export default app;
